### DISCUSSION TITLE
3.0.2に関する不具合や質問はこちら

### DISCUSSION THREAD

**Discussion by dongaba** [dongaba/TVerRec#69] [Issues]
### 3.0.2 Changed

- 不要ファイル削除処理の可読性とパフォーマンスを向上

### 3.0.2 Fixed

- Mac と Windows/Linux 混在環境で移動処理が濁点や半濁点を含むディレクトリを取りこぼす問題を修正

**Full Changelog**: https://github.com/dongaba/TVerRec/compare/v3.0.1...v3.0.2

<hr /><em>This discussion was created from the release <a href='https://github.com/dongaba/TVerRec/releases/tag/v3.0.2'>3.0.2</a>.</em>

**Comment by koyama4284** [comment]
PowerShell のバージョン

2.9.7 だったか 2.9.8 のときに、c.validate_video.cmd が戻り値の関係で
うまく動作しないことがありました。原因は PwerShell 7.4.0 の既知バグの
せいでした。
dongabaさんの方で既知バグ対応もしていただいておりますが、私みたく
 7.3.9 のまま自動更新しない設定にしている人もいるかと思います。

3.0.1 の Discussionsを読んでその事に気が付きました。dongabaさんの方は 
7.4.1 で開発されているようですので、自動更新しないにしている人も
バージョンは合わせておいた方がいいですね。

<notextile>#</notextile> 自分への備忘と啓蒙です

**Comment by dherhi** [comment]
x.generate_list.cmdを実行したところ

```
Convert-Path: E:\TVerRec\conf\user_setting.ps1:178
Line |
 178 |  … stAppLogo = Convert-Path (Join-Path $script:imgDir './TVerRec-Toast.p …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path 'E:\TVerRec\img\TVerRec-Toast.png' because it does not exist.
Convert-Path: E:\TVerRec\conf\user_setting.ps1:181
Line |
 181 |  … :iconPath = Convert-Path (Join-Path $script:imgDir './TVerRec-Icon.pn …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot find path 'E:\TVerRec\img\TVerRec-Icon.png' because it does not exist.
Write-Error: E:\TVerRec\src\generate_list.ps1:22
Line |
  22 |      . (Convert-Path (Join-Path $script:scriptRoot '../src/functions/i …
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | ❗ 外部関数ファイル(common_functions.ps1)の読み込みに失敗しました
Press any key to continue . . .


```

と出てリストが作成できなくなりました｡　

手動アプデ後､できなかったので最新のZIPからデータをもってきて上書きしてるのですが､前回作っていたlistフォルダも消えてしまいDBに統合？されたのかなと思いますが､その際に　list管理用に別のEXcelを入れていたためそれも消えてしましました｡
もしよろしければ､アプデで整理されたファイルはゴミ箱に移動もしくはOLDフォルダなどでしばらくは保持できるようにしていただけないでしょうか？
とりあえずは余計なものを実行下に置かないようにいたします｡

**Reply by dongaba** [reply]
バージョンアップ時には、最新版を個別のファイルごとに上書きする仕様としているので、既存ファイルを勝手に削除することはありません。
念の為、v2.9.9からv3.0.2へのアップデートを当方環境で確認しましたが、エラーが出るという事象も、自分で置いたファイルがなくなるという事象も確認できませんでした。

**Reply by dherhi** [reply]
検証ありがとうございます｡
こちらではリスト作成のために別にTverrecを使っており､v2.9.9から`update_tverrec.cmd`を用いてアップデートしたところ､
```
-----------------------------------------------------------------
アップデートの作業ディレクトリを削除します

-----------------------------------------------------------------
過去のバージョンで使用していたファイルを削除、または移行します
```

の段階でlist､img,colabフォルダを始め複数のフォルダが削除が確認されました｡

私の書き方が間違っていたのですが､手動アプデ後ではなく､`update_tverrec.cmd`を使う→更新後実行できなかったので手動アプデでした｡

**Reply by dongaba** [reply]
2.9.8の時点で挙げて頂いたフォルダは統廃合(listはdbと統合、その他2つはresources配下に移動)したので、2.9.8以前からアップデートされたのではないでしょうか。

**Reply by dherhi** [reply]
2.9.8以前からアップデートというのは､その通りだと思います｡
それ以前に手動または`update_tverrec.cmd`を持ちいてアップデートしておりました｡
事実としては､その過去から積み上げた2.9.9から`update_tverrec.cmd`を用いてアップデートしたところ､複数のフォルダまたはファイルが消失した､ということです｡

**Reply by dongaba** [reply]
ご報告ありがとうございました

**Reply by dherhi** [reply]
こちらこそいつもありがとうございます｡

**Comment by dherhi** [comment]
別で質問なのですが､リスト作成が追記方式ではなくなったのでしょうか？
以前は追記方式だったので新規番組をエクセルから引き出せるように
`=UNIQUE(CONCATENATE("series/", FILTER(list!$B:$B, list!$L:$L="tag/drama"), " #", FILTER(list!$A:$A, list!$L:$L="tag/drama")))`
のようなもので別シートのCSVから呼び出せるようにしていたのです｡
追記方式に設定できるようにできますでしょうか？

**Reply by dongaba** [reply]
現在も追記になっていると思います。列の数が変わったと思うのでそのせいかもしれませんね。追記されていないということは実際にファイルを開いて確認されましたでしょうか?

**Reply by dherhi** [reply]
💡 番組をリストファイルに追加します
という行が以前まででは追記の分だけ表示されていたので､以前とは違い新規に上書きしているものと考えていました｡

認識の不一致だったら申し訳ないのですが､キーワードに対しサイトより一致したものを取得し､
比較したものを最後の行に追記というのが私の今までの追記方式の認識でした｡
あっていますでしょうか？

ファイルを開いて見てはいるのですが､こちらの見方といたしましては仮に追記の方式が行に挿入やソートされているのであればわかりません｡　最終行のタイトルでしか目視判断は不可能でした｡

新しく取得したものが最後に追記されているのであれば問題ないです｡
とりあえずもう一度リスト作成したところ取得件数はフルで処理されていますが､追記分は最後に追加されているように見えます｡

**Reply by dongaba** [reply]
認識あっています。
依然として問題があるようでしたら教えて下さい。

**Reply by dherhi** [reply]
表示件数が追加分だけではなく全県だったので処理として以前と比べて遅くなっているのではないかと思ったのです｡
勘違いでしたら申し訳ないです｡　ありがとうございます｡

**Comment by cutewombat2** [comment]
こんばんは。リストファイル作成がほぼ途中で終了してしまいます。
以前使っていたversion（不明。2.8とか2.9だった記憶あり）では発生しなかったのですが、version3.0.2で頻発します（感覚として10回やって1回しか最後まで取得できない感じ）。

以下の参考画像では、２番目以外は途中で終了しちゃっております：
![2024-04-06_122259](https://github.com/dongaba/TVerRec/assets/167008693/eeafb295-f858-4bb2-9883-259d0e752835)
![2024-04-06_190814](https://github.com/dongaba/TVerRec/assets/167008693/94c9718b-8c55-4543-b8f7-ca86483e4731)
![2024-04-14_132034](https://github.com/dongaba/TVerRec/assets/167008693/56218a6f-c20d-4ddd-a5e5-497569ce044a)
![2024-04-14_152812](https://github.com/dongaba/TVerRec/assets/167008693/bc931c71-0ee6-4951-9c89-29cecc522970)
![2024-04-14_223302](https://github.com/dongaba/TVerRec/assets/167008693/85e1835f-c948-4752-aa0e-d3a69cc73b74)

keyword.confの中身：
```
#━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#ダウンロード対象キーワード
#━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
#----------------------------------------------------------------------
#	「#」でコメントアウト
#----------------------------------------------------------------------
sitemap			#配信中全番組

```

**Reply by dongaba** [reply]
試してみたところ、たしかに6件ほどエラーとなったようです。発生率は0.1%なのでHTTP接続エラーなどの可能性が高いように思いますが、エラーハンドリング等で回避できるか、1回だけリトライするようにするかなど検討してみようと思います。

```
--------------------------------------------------
5695/5697 - https://tver.jp/episodes/epzz5r9v7u
💡 番組をリストファイルに追加します
--------------------------------------------------
5696/5697 - https://tver.jp/episodes/epzze4hv7x
💡 番組をリストファイルに追加します
--------------------------------------------------
5697/5697 - https://tver.jp/episodes/epzzgjyd7t
💡 番組をリストファイルに追加します

---------------------------------------------------------------------------
番組リストファイル出力処理を終了しました。
💡 必要に応じてリストファイルを編集してダウンロード不要な番組を削除してください
　リストファイルパス: /Users/tverrec/Public/TVerRec/db/list.csv
---------------------------------------------------------------------------
Finished ...
~/Public/TVerRec/unix   master  ~10           3h 27m 52.987s  tverrec  zsh 
❯ wc -l ../db/list.csv
    5692 ../db/list.csv
~/Public/TVerRec/unix   master  ~10                     26ms  tverrec  zsh 
❯ cat ../VERSION
3.0.2%
~/Public/TVerRec/unix   master  ~10                     25ms  tverrec  zsh 
❯
```

**Reply by cutewombat2** [reply]
早朝にもかかわらず検証して頂きありがとうございます。
検証中に「途中で終了すること」は起こりませんでしたでしょうか？
こちらの環境（Windows11）では、「処理停止」ボタンを押していないのに、途中で終了し、「処理を終了しました」と表示されてしまうことが頻発します。
以前のパソコン（Windows10）で以前のversionのTVerRec（2.8か2.9の記憶あり）を使っていたときは何十回と「リスト作成」しましたが「一度も」処理が途中で終了することはありませんでした。

パソコンを使っている最中に「リスト作成」すると、早い段階でTVerRec GUIへのログ出力が滞り始め、しばらくして見ると停止していることが多いように感じられます（2000件台で止まってしまう）。
一方で、「リスト作成」自体処理件数が多いため時間のかかる処理なので、寝ている間や外出中に実行することも多いのですが、その場合は惜しいところまで行っていることが多いように感じられます（4000件台で止まっている）。

これがとても使い勝手を悪くしているので、何かが起こった場合（何が終了を引き起こしているのか不明ですが、HTTP接続エラーなどでしょうか）でも最後まで処理を継続するように修正して頂きたいです。

**Reply by dongaba** [reply]
GUI版は、エラーメッセージ等の色付けをしたため、行数が増えてくるとメモリを大量に使っているようです。ログクリアをすると多少改善するのですが、抜本的にはログの最大行数を設定する機能を追加予定です。
当方が現在テストしている限りではGUIの応答速度は大幅に改善しています。
停止するかについては、実は私はリスト作成機能を普段全く使っていないのでよくわからないのですが、先ほど試した限りは停止はしませんでした。(3.0.2)
また、再実行したところエラーとなったものも処理できたので、おそらくHTTP接続のタイムアウトなどではないかと思われます。
停止してしまう原因として考えられるのはPowerShellのプロファイルでエラー時停止などの設定をしている可能性が考えられますが、プロファイルなどで設定していないでしょうか？

**Reply by cutewombat2** [reply]
PowerShellのプロファイルでエラー時停止などの設定はしていません（TVerRecのインストール時に一緒に自動ダウンロードされ、PowerShellはTVerRecでしか使っていませんし設定もいじっていません）。
今回、設定を色々変えましたが、結果は変わりませんでした。
また、出かける前に800くらいのログ出力で「実行ログクリア」ボタンを押し、帰宅したところ、やはり途中で終了してしまっていました。
ずっと見張って、定期的に「実行ログクリア」ボタンを押し続ければ最後までいくのか、試してみます。
（しかし、これで解決できても、時間がかかる作業なので、完全放置でちゃんと最後まで動作して欲しいですね。）
![2024-04-16_220249](https://github.com/dongaba/TVerRec/assets/167008693/15fb756a-dfda-4113-9c12-dec685e0419b)


>当方が現在テストしている限りではGUIの応答速度は大幅に改善しています。

はい、以前のPC（Windows10）で以前のversionのTVerRec（2.8か2.9だった記憶あり）と比べ、かなり早くなっているのは感じます。
以前は2時間くらい放置しないと作業完了しませんでした（そのかわり途中で終了することもなかった）。
今のPCの方が高スペックなので、一概には比較できませんが、1300件あたりまでは連続でログ出力され続け、あっという間に届きます。その後は、ログ出力の頻度が鈍りだし、2000件くらいになると動いてるのか止まってるのか分からないくらいにログ出力が散発的になります（数分おきに確認すると「ログ出力が少し増えている」ので「ちゃんと動いているな」と確認できる感じです）。
これも、定期的に「実行ログクリア」ボタンを押し続けることで1300件くらいまでの快調さが続くのか、観測してみます。

>実は私はリスト作成機能を普段全く使っていないので

「ループ実行」だと取得される動画はランダムなので、endTimeが数か月先の動画、さらに言うとよりによって大量にある「endTimeが2030/03/31 23:59:59と2050/12/31 23:59:00」の動画（全動画数4000～5000件の内、これらは実に2109件も占めるので、約40～50%の確率でダウンロードされる）、体感的には「シナぷしゅ」がダウンロードされ続けてしまいます。肝心の、もうすぐ終了してしまう動画がダウンロードされることは確率的にほぼありません。
なので、TVerRecは記憶では去年の11月頃から使っていますが、試行錯誤の末、「リスト作成」でデータを取得（数日おきに実行）→endTimeの昇順に並べ替え→「リストダウンロード」という今の使い方に落ち着きました。
ですので、「リスト作成」が中途半端に終了してしまうという現状はとてもツラいです。

**Reply by koyama4284** [reply]
GUI は使っていないので本件とは関連が無いかもしれませんが、PowerShell のバージョンはいくつをお使いでしょうか?

どうも先週リリースされた 7.4.2 を使うと不具合が出るように思われます。家に帰ったら追試予定なのですが、7.4.1 までは CLI でのリスト作成(x.generate_list.cmd) → ダウンロード(y.download_list.cmd) → コンペア(c.validate_video.cmd )で問題は出ていなかったのですが、7.4.2 以降動きがちょっと変。(うまく表現できなくて済みません。)

dongabaさん、cutewombat2さんの使用している PowerShell のバージョンはいくつになりますか?
<notextile>#</notextile> 自動更新しないからするに変えたけど、場合によっては 7.4.1 で自動更新しないの方がいいかも。

**Reply by koyama4284** [reply]
家に帰って追試しましたが、PowerShell のバージョンは関係なかったです。
以前 c.validate_video.cmd  がうまく動かなかったので原因かもと思ったのですが、的外れでした...

お騒がせしました。

**Reply by dongaba** [reply]
追加の情報提供ありがとうございました。
また何かあればよろしくお願いします。

**Reply by dongaba** [reply]
@cutewombat2 
まだリリース版として登録していませんが、v3.1.0を用意しました。
以下からダウンロードいただけるので、差し支えなければ試してみて問題が解決しているか教えていただけませんか。
念の為、現在の3.0.2とは別のフォルダにダウンロードしてお使いください。(設定項目が増えているので、設定ファイルをコピーした際は新規設定項目はデフォルト値で動作します)
![image](https://github.com/dongaba/TVerRec/assets/83079591/1b9dc60c-526d-49b3-9505-af3b36b92ac1)

また、リスト作成処理を並列処理するようにしたので、処理時間もだいぶ短縮するはずです。(ざっくりですが、デフォルト値で
並列動作させた場合1/5位になるかと思います)

特に問題がなさそうであれば、近日中にリリース版として登録しようかと思います。

**Reply by dongaba** [reply]
3.1.0での変更点は書ききれませんが、ざっと以下にまとめてあります。
https://github.com/dongaba/TVerRec/blob/master/CHANGELOG.MD#310
